File: Leetcode/Python/_120.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        for i in range(1, len(triangle)):
            for j in range(len(triangle[i])):
                if j == 0:
                    triangle[i][j] = triangle[i][j] + triangle[i - 1][j]
                elif j == len(triangle[i]) - 1:
                    triangle[i][j] = triangle[i][j] + triangle[i - 1][j - 1]
                else:
                    triangle[i][j] = min(triangle[i - 1][j - 1] + triangle[i][j], triangle[i - 1][j] + triangle[i][j])
        
        return min(triangle[-1])


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        for i in range(len(triangle) - 2, -1, -1):
            for j in range(len(triangle[i])):
                triangle[i][j] = min(triangle[i + 1][j] + triangle[i][j], triangle[i + 1][j + 1] + triangle[i][j])
        
        return triangle[0][0]


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        ROWS = len(triangle)
        dp = [[inf for _ in range(cols)] for cols in range(1, ROWS + 1)]
        dp[0] = triangle[0]

        for row in range(1, ROWS):
            for col in range(len(triangle[row])):
                right = dp[row - 1][col] if col < len(triangle[row - 1]) else inf
                left = dp[row - 1][col - 1] if col - 1 >= 0 else inf

                dp[row][col] = min(right + triangle[row][col], left + triangle[row][col])

        return min(dp[-1])


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        ROWS = len(triangle)
        prev = triangle[0]
        
        for row in range(1, ROWS):
            cur = triangle[row]
            for col in range(len(triangle[row])):
                right = prev[col] if col < len(triangle[row - 1]) else inf
                left = prev[col - 1] if col - 1 >= 0 else inf

                cur[col] = min(right + triangle[row][col], left + triangle[row][col])
            prev = cur
        return min(prev)

File: Leetcode/cpp/120.cpp
class Solution {
public:
  int minimumTotal(std::vector<std::vector<int>>& triangle) {
    int m = static_cast<int>(triangle.size());
    if (m <= 1) {
      return triangle[0][0];
    }
    int n = static_cast<int>(triangle[0].size());

    int output = INT_MAX;
    for (int i = 1; i < m; ++i) {
      n = static_cast<int>(triangle[i].size());
      for (int j = 0; j < n; ++j) {
        if (j >= 1 && j < n - 1){
          triangle[i][j] = std::min(
            triangle[i - 1][j - 1] + triangle[i][j],
            triangle[i - 1][j] + triangle[i][j]
          );
        } else if (j == 0) {
          triangle[i][j] = triangle[i - 1][j] + triangle[i][j];
        } else {
          triangle[i][j] = triangle[i - 1][j - 1] + triangle[i][j];
        }
      }
      if (i == m - 1) {
        for (int j = 0; j < n; ++j) {
          output = std::min(output, triangle[i][j]);
        }
      }
    }
    return output;  
  }
};
